kaldb/src/main/java/com/slack/kaldb/logstore/search/KaldbDistributedQueryService.java

```diff
@@ -126,6 +128,42 @@ private KaldbServiceGrpc.KaldbServiceFutureStub getKaldbServiceGrpcClient(String
         .withCompression("gzip");
   }
 
+  private Set<KaldbServiceGrpc.KaldbServiceFutureStub> getSnapshotsToSearch(
+      long startTimeMs, long endTimeMs) {
+    Set<String> snapshotsToSearch =
+        snapshotMetadataStore
+            .getCached()
+            .stream()
+            .filter(
+                snapshotMetadata ->
+                    containsDataInTimeRange(
+                        snapshotMetadata.startTimeEpochMs,
+                        snapshotMetadata.endTimeEpochMs,
+                        startTimeMs,
+                        endTimeMs))
+            .map(snapshotMetadata -> snapshotMetadata.name)
+            .collect(Collectors.toSet());
+
+    return searchMetadataStore
+        .getCached()
+        .stream()
+        .filter(searchMetadata -> snapshotsToSearch.contains(searchMetadata.snapshotName))
```

Comment:
If a snapshot has multiple replicas, would we query both replicas in this case?

How do we handle the case when the snapshot exists but is not present in the cache?